### ISSUE TITLE
shape data migration results cause request_stream GstElement error

### ISSUE THREAD

**Issue by hburgund** [probabble/roundware-server#6]
Steps to reproduce:
1. Spin up a fresh install on `develop` branch
2. Upgrade to postgres
3. Install fixture
4. Run the data migration

The shapes appear to be fine in the admin maps, but when I try to create a stream, I get an error that I’ve never seen before:

```
  File "/vagrant/roundware/settings/../../roundwared/rwstreamd.py", line 54, in start_stream stream.start()
  File "/vagrant/roundwared/stream.py", line 70, in start self.add_speakers()
  File "/vagrant/roundwared/stream.py", line 159, in add_speakers speakers)
  File "/vagrant/roundwared/gpsmixer.py", line 73, in __init__self.move_listener(listener)
  File "/vagrant/roundwared/gpsmixer.py", line 120, in move_listener self.remove(src_to_remove)
TypeError: argument must be a GstElement
```

I’ve tried several times and am getting it consistently.

However, if I don’t run the data migration and simply create a `shape` using the admin, I do not get this error and the stream is created with no issue. But if I run the data migration and then “undo” it by making another random `shape` using the admin, I still get this error. In other words, it seems that once the data conversion has been completed, simply re-making the `shape` doesn’t fix the problem. 

However, if I manually set the `boundary`, `shape` and `attenuation_border` fields back to NULL and then create a new `shape` using the admin, everything works again.

It takes only the data migration (i.e. no code changes) to get this error to happen from a previously working system, so I'm pretty confident this is where the problem lies.

It seems that there is something different about the format of the shape data between making it in the admin and doing the SQL data migration? POLYGON vs MULTIPOLYGON probably isn’t it, but maybe something along those lines? geometry, geography, etc etc?!? The database doesn’t complain about the migrated shapes, so it doesn’t violate whatever checks and constraints are set up there.

**Comment by hburgund** [comment]
email from @probabble "My instinct is that the migration I made for you is incomplete. We need to either 1) re-save the speaker objects to get the `attenuation_border` and `boundary` fields to calculate after running our custom sql or 2) calculate those fields as part of the custom sql running in the migration. Not 100% sure that will fix your problem but it's what I will attempt first."

---

I think you are definitely correct that some kind of saving/re-saving has to happen to make sure that the boundary and attenuation border fields are generated and saved themselves during the data migration process. But even if I do this manually in the admin, without touching the `shape` field, I still get the GstElement error. If I don’t do this, I get the `AttributeError: 'NoneType' object has no attribute ‘srid'` error.

I tried something new, inspired by your comment, which was to edit the `shape` generated by the data migration rather than setting it to `NULL` and creating an entirely new shape in the admin. But this didn’t work either. I also tried adding a new polygon to the circle and that failed as well in the same way (i.e. GstElement). I then tried to delete the original shape using the admin and only have the additional polygon as part of the multi polygon. This approach generated another django error that I had before:

```
Polygon must have at least four points in each ring
```

This makes me think that there is some trace of the original circle still in the shape somewhere that doesn’t have four points or something? fwiw, if I look in the db in this state, the `attenuation_border` is a very short value, way shorter than the other two, which is odd (i.e. `0103000020E610000000000000`). Could this be where the points are missing? (I think that answer is yes, but I’m not sure that helps our other issue…)

In any case, if I manually set `shape` and `attenuation_border` to `NULL`, then create a new speaker `shape` using the admin, everything works fine again. It’s as though the field needs to be entirely reset to `NULL` or else there is some vestige of the data migration process that remains and causes this issue.

**Comment by probabble** [comment]
Okay... after closing issue #5 by generating the derivative geometries via the migration, I am not able to reproduce these errors. @hburgund, would you see if that also resolves it for you?

**Comment by hburgund** [comment]
well, that would be great, and totally plausible, that your #5 changes fixed this as well. I gotta get the kids now, but will test in about an hour or so and let you know...thx!

did you attempt to reproduce the errors prior to integrating the fix for #5?

**Comment by probabble** [comment]
I didn't reproduce them prior to the fix... sorry.

**Comment by hburgund** [comment]
This issue seems to be directly related to #2 in that the same workaround of setting `minvolume` to non-zero values solves it.

I set the minimum volume to a non-zero value in order to get around #2 and once I did this, I was unable to generate the `GstElement` error any more even in situations where I consistently got it before i.e. right after a data migration without doing the whole setting to `NULL` and re-creating thing.

I also did numerous tests to see if I could generate the `GstElement` error while still doing the setting to `NULL` and re-creating in the admin and, as we suspected, if I do this and create the new speaker range such that it doesn’t cover Null Island, I still get the `GstElement` error. 

So long story short, the issue now appears to have nothing to do with the shape field directly. It looks like my entire line of inquiry/debugging was off because of a non-rigorous approach. Ugh.

The problem arises when a `request_stream` call is sent with no lat/lon in a situation where the speaker range doesn’t include 0,0 AND the `minvolume` for the speaker is 0. In other words, when no lat/lon is included in the call, it defaults to 0,0 and if 0,0 is out of range and `speaker.minvolume` is set to 0, the error happens. I haven’t done enough testing yet, but my inclination is that it has nothing to do with 0,0 specifically, but rather if the request is for an out of range location anywhere, the error will happen unless the `minvolume` is set to a non-zero value.

I hope that once we fix #2 and ensure that the data migration calculates all three gis fields, we will be good to go. We may not have to do any of the rigamarole with requiring lat/lon in geo-listen-enabled projects, though that might not be a bad idea eventually. I have a fear that the clients would need to be re-written to handle this, so I don’t want to do it right away if we can help it.
